### DISCUSSION TITLE
can not inspect database from App Inspector on Android Studio when using Drift

### DISCUSSION THREAD

**Discussion by salauddin23shumon** [simolus3/drift#1818] [Show and tell]
App Inspector is a very much helpful tools which comes with Android Studio to inspect SQLite database. 
When using ROOM, SQLite or floor lib for database, I can able to monitor my database from app inspector (even moor older version of drift). But when using new Drift, inspector shows nothing.
After investigating I have come to know that, when using -> dbFolder = await getApplicationDocumentsDirectory() and return NativeDatabase(file) for creating database, no Databases folder was created in the app directory of  mobile/device. Database file (for my example db.sqlite) was created inside app_flutter folder, hence app_inspector was not show anything. 

![withBug](https://user-images.githubusercontent.com/58369122/164983649-33e0bbaf-0bbe-40a0-9a97-85eda6003525.jpg)

Then I follow the moor database structure and database creation and found SqfliteQueryExecutor.inDatabaseFolder. For that we also need 
drift_sqflite: ^1.0.0 (older version of drift) package. Now I just check platform for Android  then return SqfliteQueryExecutor.inDatabaseFolder(path: DB name);
By doing this now Databases folder is created inside app directory of the device and database file also created inside the Databases folder.

![withoutBug](https://user-images.githubusercontent.com/58369122/164983663-a2ebcb67-d381-4eb1-8c8f-dac99ec0cd78.jpg)

Now i can inspect my app database by using app_inspector when using new Drift package.

**Comment by simolus3** [comment]
I've moved this issue into a discussion because that might improve visibility :)

As an alternative to `drift_sqflite`, you can also import `sqflite` directly but only use it to determine the database folder:

```dart
import 'package:sqflite/sqflite.dart' show getDatabasesPath;

final database = LazyDatabase(() async {
  final dbFolder = await getDatabasesPath();
  final file = File(p.join(dbFolder, 'db.sqlite'));
  return NativeDatabase(file);
})
```

Then you're still guaranteed to get the latest sqlite3 version across all devices but it should show up in the builtin database explorer.

Alternatively, you can also use the [storage_inspector](https://pub.dev/packages/storage_inspector) IDEA plugin to view and inspect drift databases easily.

**Reply by pbayog00** [reply]
I'm trying to use the App Inspection, but I'm facing problems with the connection with the database. I'm using this code with sqflite, but when I try to run the Database Inspector my db.sqlite database appears closed, so I can't inspect anything. 
The "Keep database connections open" neither works for me.
Any clue on this?
Thanks in advance
![image](https://user-images.githubusercontent.com/47598318/165482010-ee2dd071-45e3-478e-97fe-4fd0a0e7fafd.png)

**Reply by simolus3** [reply]
Looks like I was mistaken :( I thought it would pick up the database if it exists in the right folder, but apparently it also needs to be opened.

There could be a setup of opening a sqflite-based database in addition to the `NativeDatabase` just so that the database inspector picks it up. But in that case one might as well use `drift_sqflite` as a main data source.

**Reply by pbayog00** [reply]
I changed the _openConnection() code by the one given by @salauddin23shumon and it is working for me now.
Thank you both for yout help.

**Comment by Abdktefane** [comment]
i have the same issue is there any problem or drawback with using `drift_sqflite: ^1.0.0` and `SqfliteQueryExecutor.inDatabaseFolder(path: kDbFileName)` ?

**Reply by simolus3** [reply]
sqflite uses the sqlite3 library from the operating system which may be outdated or not support all extensions like json1 or fts5. A `NativeDatabase` with `sqlite3_flutter_libs` ships the latest sqlite3 with your app and uses that one.

**Reply by Abdktefane** [reply]
i got you i'll use sqflite executer only for debugging and NativeDatabase for release
thanks a lot

**Comment by ghost** [comment]
Sorry, but I don't get it :-(

I use the [simple isolates setup](https://drift.simonbinder.eu/docs/advanced-features/isolates/#simple-setup) with a `NativeDatabase `, but the connection is still closed.

Do I need this `SqfliteQueryExecutor.inDatabaseFolder(path: kDbFileName)`, or should it also work with the `NativeDatabase`?

I can't use the `SqfliteQueryExecutor` thing because then I get an older SQLite version that doesn't have all I need (SQLite JSON for example).

**Reply by simolus3** [reply]
It looks like the database needs to be opened with native Android APIs, which would require a `SqfliteDatabase` in drift.

I think you could keep using a `NativeDatabase` but, in debug builds, open a sqflite database on the same file as well. It doesn't do any harm since the sqflite database is essentially unused, but hopefully Android Studio will then see that database is being opened.

**Comment by BetterB0y** [comment]
Any news on this? @simolus3

**Reply by simolus3** [reply]
There's not much that drift could do about this. You need to open a database using Android APIs for it to become visible in the inspector by Android studio. See e.g. [this](https://github.com/simolus3/drift/discussions/1818#discussioncomment-4324225).
We also have [a DevTools extension](https://drift.simonbinder.eu/Tools/devtools) that provides similar functionality if you don't want to add a sqflite dependency.

**Reply by BetterB0y** [reply]
I'm migrating from floor now and I have a couple of questions.
- Is drift_sqflite like using floor with sqflite?
- Using drift_sqflite has some disadvantages?
- When using drift_sqflite i don't need drift_flutter?

**Reply by simolus3** [reply]
> Is drift_sqflite like using floor with sqflite?

I'm not sure if floor was doing anything special around sqflite (drift doesn't do much except wrapping sqflite for its own interfaces). So I can't say for sure, but it should be very similar in behavior.

> When using drift_sqflite i don't need drift_flutter?

Correct.

> Using drift_sqflite has some disadvantages?

I see the FFI-based interfaces as the primary method that I'm recommending to users. That doesn't mean that there's anything wrong with `drift_sqflite`, but using `drift_flutter` guarantees a recent SQLite version (sqflite uses the one from the OS) and the FFI-based interop is faster than the platform channels used by sqflite. That doesn't make much of a difference for most apps though, and if `drift_sqflite` works for your app there's nothing wrong with that.
